### PR TITLE
Sockets - Nara

### PR DESCRIPTION
**Pull request by bonara**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom model method for finding the top ten votes in my work model. The custom method is currently making a lot of database calls, I would like work on improving the method to decrease the number of calls. 
Describe how you approached testing that model method. What edge cases did you come up with? | Test cases include returning works, returning 10 works even if there are no votes.  
What are `session` and `flash`? What is the difference between them? | Flash only lasts one more request, while session last as long as the cookie does not expire or session is terminated. 
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on has_many through and has_many relationships. Also I have a better understanding on where files and folders are located in rails.
What is the Heroku URL of your deployed application? | https://nara-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | More examples and test activities on models in class would be helpful.

### CODE REVIEW

File: app/views/works/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<h2>List of Works</h2>
+<section class="media-table" >
+<% @categories.each do |category| %>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:3]: nice work DRYing this up!

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    it 'responds with 200 OK for a logged-in user' do
+      # Arrange
+      perform_login(test)
+
+      # Act
+      get current_user_path
+
+      # Assert
+      must_respond_with :ok
+    end
```
Comment (dHelmgren) [inline test/controllers/users_controller_test.rb:14]: I'd like to see more tests for login flow here. How do you make sure that when you log in using the same name that you are getting the same user?

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+describe User do
+  describe 'validations' do
+    it 'requires a username' do
+      user = User.new
+      user.valid?.must_equal false
+      user.errors.messages.must_include :username
+    end
+
+    it 'rejects a user with the same username' do
+      user = User.new(username: 'sansastark')
```
Comment (dHelmgren) [inline test/models/user_test.rb:12]: NERD! :P

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good
Comprehension questions | good
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Views are well-organized (DRY, use of semantic HTML, use of partials) | yes
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | yes
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | see comments
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Great job overall! Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
